### ISSUE TITLE
[Bug] tooltip‘s formatter property format wrong value

### ISSUE THREAD

**Issue by 848555155** [apache/echarts#17749]
### Version

5.4.0

### Steps to Reproduce

here is a simple reproduce echartsOptions：`
{
  legend: {},
  tooltip: {
    formatter: '{b}:{c} {d}% {@product} {@[0]}'
  },
  dataset: {
    dimensions: ['product', '2015', '2016', '2017'],
    source: [
      { product: 'Matcha Latte', 2015: 43.3, 2016: 85.8, 2017: 93.7 },
      { product: 'Milk Tea', 2015: 83.1, 2016: 73.4, 2017: 55.1 },
      { product: 'Cheese Cocoa', 2015: 86.4, 2016: 65.2, 2017: 82.5 },
      { product: 'Walnut Brownie', 2015: 72.4, 2016: 53.9, 2017: 39.1 }
    ]
  },
  series: [{ type: 'pie' }]
}`

### Current Behavior

![image](https://user-images.githubusercontent.com/34100171/195047540-10fcb32d-5a58-49a2-a927-e2926297fd64.png)

### Expected Behavior

{c} can display exactly value in tooltip

### Environment

```markdown
- OS: Windows 11
- Browser: Microsoft Edge 106.0.1370.37
- Framework: nomatter
```

**Comment by 848555155** [comment]
> The dataset items are objects and that's why format '{c}' shows an object placeholder. Did you get the dataset from [here](https://echarts.apache.org/en/option.html#series-pie.tooltip.formatter)? I think _dimensions_ is misleading. Please see the dataset in [this example](https://echarts.apache.org/examples/en/editor.html?c=map-usa-projection&code=PYBwLglsB2AEC8sDesBQtZmMANpEAXMrAGbABOAtgIZhgCm5RA5EgEYC-RSAxh8gBMOAUmawOAGjQYBtagGd6YbtIywA9ACpYAiJXrR5UQ0QDazEOWACArjzDMpzAEwAGAIwBWR7BceAbD5-7gDszAC6Uuhq8sA25Dz0ZqpqxJbWdsq-ALK0PAAW1LAAMrQMPm5eRAAsAMwAdLVSlf5EABye9W3NHiFEAJwNIeJRqRgo6bb2LNkQOADWsAAq9NQVHp7tDe497q2wIQ3Vu32wnp3uIylqE1ZTWcwAwvn09Iqwj8A8wGu7m7Btfz1Y6wFpEfydZwndrOeqeK7RVK3DLTXwAdWoOGgNjAsAAQlYAO7QCD0dZVA6wkFgs4NfrQ2C1fr1S4cRGwcKwTTqdns3T6QzGeRmCx3TJBSreJwtCW9CJSDTqWAUWAQQUCeiwWLxRLs7UJJKwUzsjDqdTmSbi6UbWV7W1hSImo3MXJgApFUp0MlSOqNKQdLpSQb1EKOsbO2YLZarHxtbZSQ7AqTnFlhsbmZ6vd6fb6_AFAkEQ-pQgGwzxp1LmDFYnH4okk72UpO0-r0xnM9zhdldjCSVSKcik4VGlBgACeIENFlJPhw1DY9Bw3DIVDKjBYSAAApb7PwAITMfgcTlsjCUMfPajkMD1RRgADy4GMAApQJAYABKADcQA). It works with format '{c}'. And formats {@Product} {@[0]} are for [labels](https://echarts.apache.org/en/option.html#series-line.label.formatter) only. ![image](https://user-images.githubusercontent.com/13038071/195170798-be537702-d9ab-44a0-b88a-b52ac4754da6.png)

It works with format '{c}', but it shows all of datas in this item.It seems that echarts format '{c}' using Array ['Milk Tea', 83.1, 73.4, 55.1]. 
When using series.data to display, format '{c}' only shows item’s value. I think format behavior in dataset should be consistent with in series.data.
![image](https://user-images.githubusercontent.com/34100171/195229264-456814a1-2e6e-485f-9855-64e4594a8bcb.png)
here is a  [demo](https://echarts.apache.org/examples/en/editor.html?c=pie-simple&code=PYBwLglsB2AEC8sDeAoWtJgDYFMBcya6GOAHmAQOQBKOAZjgE5OzB2wCGsA6jgEYBnCGByUANEXQCArnxHkqAMQ4BrHLAAiHMB3GTYuOhViUAxjmgjGlIgF8J6MMGBZIIAqmIZGEAOa-mKmEcAFs9LzpgRhDtKyokPlsPU1tkABNbAFIbdHsiXADoNI99KIgLY0oANyZIUw4scOJDSpac2DypJnKBAgBtfU8vWGgOEPwTAEFTcwEBWEVGYDCHYbAATxAJyhBypq9GDjSIaV6TAFYABmzVrzTtDn79LyRYKobpCYBGS4AWAA4xCMxtsAMo4DiMUwAC1gAFFoL4INBRB1bsNXu8sJ8CAB2ADM5yBo3GVA0EGYpjAlDRz2ImI-E3O_0uxJBVDhMQgjVpw3pb0ZBABvzZpJMAFVoFA4JM0gIaZ1hgBdfS2Igq2wAbiAA) that shows format '{c}' behavior  in series data.

**Comment by helgasoft** [comment]
@848555155, agree - _{c}_ format is **not consistent** across data sources.
It would be preferable in my example tooltip to show **Milk Tea, 83.1, 29.14.%**  where 83.1 is the value taken to build the pie slice.

BTW, _valueFormatter_ shows the correct data value:   ``` valueFormatter: (value) => value```
but cannot be combined with other formats like _{b}_...

**Comment by Tracts81** [comment]
I also meet this problem now. I don't know if there are similar issues before, but just find here.
I have tried ways you mentioned and also found it no effective. Besides I configured the series.encode.itemName and series.encode.value and got the same result.
So I turn to used the way of callback function and found: The 'value' from parameter is also returns an Array in the callback function when using dataset, instead of a Number using series data.
Though I resolve it by callback function, I want to know why the value is not just a number when using dataset?
